### ISSUE TITLE
Multi language

### ISSUE THREAD

**Issue by jbelien** [thomersch/openstreetmap-calendar#22]
Hello, 

As discussed yesterday, I think it's indeed a excellent idea to integrate your feed to our website (see https://github.com/osmbe/website/issues/92).
One specificity of our events is that we try to publish our events in 3 languages (EN + FR + NL).
Do you see an convenient way to do this on https://osmcal.org ?
I of course don't mind creating 3 events but if the 3 events show up on the main page, it will be really messy, really fast.

Thanks :)

**Comment by thomersch** [comment]
Good point, I already struggled around with [Kölner Hack Weekend](https://osmcal.org/event/205/), because I wanted both DE and EN, so I copied the idea from Karlsruhe by interleaving the two languages paragraph-wise.

We could also offer to add multiple language versions of the description, but we'd need to build some non-confusing user interface. What could work for the moment is to have all (three) versions in the description text and link them with anchors. I think this could work using the markdown functionality that's already there.

**Comment by qeef** [comment]
I will write my proposals as list:
- Use the default description language the language of the country of the event.
- Let users/attendees translate the event description (I used this approach in https://client.damn-project.org/).
- Make tabs for language switch (I used this approach in https://qeef.gitlab.io/mapathoner/).
- I would store the translations in database as JSON object. My mind pops up two options -- create new JSON column to event called *translations*, or change *description* column from `text` to `JSON`.

**Comment by jbelien** [comment]
> Use the default description language the language of the country of the event.

That doesn't work for Belgium and that's the reason I opened this issue !
All our events are (at least) bilingual (French + Dutch).

**Comment by qeef** [comment]
> > Use the default description language the language of the country of the event.
> 
> That doesn't work for Belgium and that's the reason I opened this issue !
> All our events are (at least) bilingual (French + Dutch).

I meant the default language, not the only one.

I think that if a description is long, it would be messy using just anchors (I mean -- multiple paragraphs in one language, then in the second, third, ...). Therefore, I thought about hiding not selected languages.

The switching between languages (does not depend if hidden or just anchored below) would be done by some navigation panel just below the date.

Then, the question about how to find out the list of languages arises -- when only text is stored in the database, I think that parsing headers (i.e. header level 1) and generating is inevitable. If JSON dictionary is stored in the database, the key would be language code and the value the description.

**Comment by thomersch** [comment]
> I think that if a description is long, it would be messy using just anchors

This is true, but it’s a solution which doesn’t break anything and doesn’t need any code.

This is not a data schema problem, but a variety of challenges: There is no such thing as a default language for many countries. Also, some regions will prefer an entirely different language, e.g. in Germany some things would be in English eventhough it’s not an official language.

You‘re talking about navigation, @qeef, but this is not the only way of consuming osmcal, there are also feeds and the API. So, if we‘re talking about solutions, we need to dig deeper. This is not a simple issue, and will not be solved through changing data types ;)

**Comment by qeef** [comment]
> > I think that if a description is long, it would be messy using just anchors
> 
> This is true, but it’s a solution which doesn’t break anything and doesn’t need any code.

Then it probably makes sense to start with this solution.

> This is not a data schema problem, but a variety of challenges: There is no such thing as a default language for many countries. Also, some regions will prefer an entirely different language, e.g. in Germany some things would be in English eventhough it’s not an official language.

If I understand correctly, this would mean that there should be the selectable default language for the event?

> You‘re talking about navigation, @qeef, but this is not the only way of consuming osmcal, there are also feeds and the API. So, if we‘re talking about solutions, we need to dig deeper. This is not a simple issue, and will not be solved through changing data types ;)

I think that if the description should be structured by languages, it can't be in one text field and data type would need to be changed. If not and there will be just convention about how to write a description (i.e. for multilingual, use header level 1 for language name), the description data type can be text.

I don't see any other solution (but maybe there is?) that avoids adding tables/columns to the database.

**Comment by thomersch** [comment]
The core issue I see here is the complexity we would be adding, if we decide to solve this problem beyond a description convention. I do see that some communities or some events need multiple languages, but we need a solution that is ergonomic to event creators, consumers of the website and the feeds and developers which use the API. If we add something in the base handling of events, we need to have all of those in mind.

Having more data structures is nice for us developers, but in the end I think we have a tradeoff here between usability and correctness.

> If I understand correctly, this would mean that there should be the selectable default language for the event?

Not really, e.g. what default language would the Belgians choose? It's basically undecidable. We would need to go by user language. But for this we need the users to be able to override their browser language etc.

> I don't see any other solution (but maybe there is?) that avoids adding tables/columns to the database.

Sure, if we decide to make multi-language descriptions happen, adding tables will be necessary. I don't mind additional tables or fields. But this opens other questions: Shall we internationalize the website first? Would we need to have event titles in multiple languages? How about other calendar systems (first day of week, …)?

I do see the benefit of having this feature implemented properly and I do have some rough ideas how to tackle this, but considering the substantial effort, I am not jumping immediately onto that, since we have some rough ways how to circumvent this issue for the moment.
